lib/services/elasticsearch.js

```diff
@@ -1360,218 +1241,213 @@ class ElasticSearch extends Service {
+    collection,
+    documents,
+    { refresh=null, timeout=null, userId=null } = {})
+  {
     const
-      index = request.input.resource.index,
-      type = request.input.resource.collection,
-      toImport = [],
-      extracted = extractMDocuments(request, {
+      esIndex = this._getESIndex(index, collection),
+      kuzzleMeta = {
         _kuzzle_info: {
-          active: true,
-          author: this._getUserId(request),
+          author: String(userId),
           updater: null,
           updatedAt: null,
-          deletedAt: null,
           createdAt: Date.now()
         }
-      });
+      },
+      {
+        rejected,
+        extractedDocuments,
+        documentsToGet
+      } = this._extractMDocuments(documents, kuzzleMeta, { prepareMGet: true });
+
+    return this.client.mget({ index: esIndex, body: { docs: documentsToGet }})
+      .then(({ body }) => {
+        const
+          existingDocuments = body.docs,
+          esRequest = {
+            index: esIndex,
+            body: [],
+            refresh,
+            timeout
+          },
+          toImport = [];
 
-    return this.client.mget(extracted.mgetRequest)
-      .then(mgetResult => {
-        const esRequest = initESRequest(request,
-          ['consistency', 'refresh', 'timeout', 'fields']);
         let idx = 0;
 
-        esRequest.body = [];
-
-        for (let i = 0; i < extracted.documents.length; i++) {
-          const document = extracted.documents[i];
-
-          if (typeof document._id === 'string') {
-            // document._id should always be equal to mgetResult.docs[idx]._id
-            // if the document is not found in ES, then we reject it
-            if (mgetResult.docs[idx].found) {
-              esRequest.body.push(
-                {index:
-                  {_index: index,
-                    _type: type,
-                    _id: document._id}
-                });
-              esRequest.body.push(document._source);
-              toImport.push(document);
-            } else {
-              extracted.rejected.push(
-                {document,
-                  reason: 'cannot replace a non-existing document (use mCreateOrReplace if you need to create non-existing documents)'});
-            }
+        for (let i = 0; i < extractedDocuments.length; i++) {
+          const document = extractedDocuments[i];
+
+          // Documents are retrieved in the same order than we got them from user
+          if (typeof document._id === 'string'
+            && existingDocuments[idx].found
+          ) {
+            esRequest.body.push({
+              index: {
+                _index: esIndex,
+                _id: document._id
+              }
+            });
+            esRequest.body.push(document._source);
+
+            toImport.push(document);
+          } else if (typeof document._id === 'string'
+            && ! existingDocuments[idx].found
+          ) {
+            rejected.push({
+              document: document,
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Use shorthand for property "document". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS3498)